### PR TITLE
Add express-validation middleware

### PR DESCRIPTION
**Pull request by eraviolo**
This PR is to test/review express-validation middleware:
- Added middleware/validation folder with: validateRequest function and schema to validate getForest request params  
- Chained validation call in /server/api/geo/index.ts, prior to gee auth. middleware call   
- Changed ForestLayerRequest params that are not string to any to be able to used directly the req sanitized values (e.g string params converted to numbers or booleans)

### CODE REVIEW

File: src/server/middleware/validation/geoSchemes.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+import { ForestSource, precalForestAgreementSources } from '@meta/geo'
+
+export const forestLayerSchema = [
+  query('countryIso')
+    .exists()
+    .withMessage('Country ISO is required')
+    .isLength({ min: 3, max: 3 })
+    .withMessage('Country ISO must be a 3 char value')
+    .bail(),
```
Comment (eraviolo) [inline src/server/middleware/validation/geoSchemes.ts]: I have to add here a validation against CountryIso type valid options

File: src/server/middleware/validation/index.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  return async (req: Request, res: Response, next: NextFunction) => {
+    await Promise.all(checks.map((check: ValidationChain) => check.run(req)))
+
+    const result = validationResult(req)
+    if (result.isEmpty()) {
+      return next()
+    }
+
+    const errors = result.array()
+    return res.send(errors)
```
Comment (minotogna) [inline src/server/middleware/validation/index.ts]: this validation middleware you added is really well done 👍.
before sending error back,I would just add a 400 status code. what do you think?

Comment (eraviolo) [reply src/server/middleware/validation/index.ts]: Yes, I will add the code before sending errors back

File: src/meta/api/request/geo/layer.ts [hunk 3]
```diff
@@ -10,9 +10,9 @@ export type ForestLayerRequest = Request<
   {
     countryIso: CountryIso
     forestSource: ForestSource
-    gteHansenTreeCoverPerc?: string
-    onlyProtected: string
-    opacity: string
+    gteHansenTreeCoverPerc?: any
+    onlyProtected: any
+    opacity: any
```
Comment (sorja) [inline src/meta/api/request/geo/layer.ts:15]: why was this changed to any?

File: src/server/controller/geo/getForestLayer.ts [hunk 4]
```diff
@@ -27,7 +27,7 @@ export const getForestLayer = async (props: Props): Promise<Layer> => {
       .getMap(
         {
           palette: metadata.palette,
-          opacity: opacity === undefined ? 1 : Number(opacity),
+          opacity: opacity === undefined ? 1 : opacity,
```
Comment (sorja) [inline src/server/controller/geo/getForestLayer.ts]: can this be
`opacity: opacity ?? 1`

File: src/server/middleware/validation/geoSchemes.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    .isIn(precalForestAgreementSources)
+    .withMessage('Not valid forest source'),
+
+  query('gteHansenTreeCoverPerc')
+    .if(query('forestSource').equals(ForestSource.Hansen))
+    .exists()
+    .withMessage('Hansen tree cover percentage is required')
+    .bail()
+    .isInt({ min: 0, max: 100 })
+    .withMessage('Not valid Hansen tree cover percentage 0-100')
```
Comment (sorja) [inline src/server/middleware/validation/geoSchemes.ts]: Can we add new translation file 'validations.js' and import it in 'en.js' and add these there?

File: src/server/middleware/validation/index.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { NextFunction, Request, Response } from 'express'
+import { ValidationChain, validationResult } from 'express-validator'
+
+const validateRequest = (checks: Array<ValidationChain>) => {
```
Comment (sorja) [inline src/server/middleware/validation/index.ts:4]: This looks good, good job 👍

File: src/meta/api/request/geo/layer.ts (hunk 3 above)
Comment (eraviolo) [reply src/meta/api/request/geo/layer.ts:15]: When request validation is performed, I convert some query params  to number. I wanted to use directly those sanitized values, but request params cannot be typed with other than string type as I understand. With any I can pass converted params (as numbers) to the controller function, maybe there is a better workaround.

File: src/server/controller/geo/getForestLayer.ts (hunk 4 above)
Comment (eraviolo) [reply src/server/controller/geo/getForestLayer.ts]: Sure, I was not familiar with that operator :)

File: src/server/middleware/validation/geoSchemes.ts (hunk 5 above)
Comment (eraviolo) [reply src/server/middleware/validation/geoSchemes.ts]: Yes, I will change this, I wanted to ask you about how to handle localized messages.

Comment (sorja) [reply src/server/middleware/validation/geoSchemes.ts]: Normally we send translation key to frontend where we use `{ t } = useTranslation()` and later in the component `t('translation.key')`
Rarely on server side we have to initialise i18n and use it there, but there are some cases (emails)

File: src/server/middleware/validation/index.ts (hunk 2 above)
Comment (eraviolo) [reply src/server/middleware/validation/index.ts]: I removed this control in last commit because I am setting default values when validating/sanitizing request params

File: src/server/middleware/validation/geoSchemes.ts (hunk 5 above)
Comment (eraviolo) [reply src/server/middleware/validation/geoSchemes.ts]: As we are using this api only internally maybe we can keep it simple for now by no overriding the default express-validator error message, what do you think? Default error message is "Invalid value" with param name and value, something like:
{
    "value": "3",
    "msg": "Invalid value",
    "param": "opacity",
    "location": "query"
  }

File: src/server/controller/geo/getForestLayer.ts (hunk 4 above)
Comment (eraviolo) [reply src/server/controller/geo/getForestLayer.ts]: I removed this control in last commit because I am setting default values when validating/sanitizing request params

File: src/server/middleware/validation/geoSchemes.ts (hunk 5 above)
Comment (sorja) [reply src/server/middleware/validation/geoSchemes.ts]: Looks good

Comment (minotogna) [reply src/server/middleware/validation/geoSchemes.ts]: as msg pass a msgKey . like msgKey:'validation.errors.invalidValues'

File: src/meta/api/request/geo/layer.ts (hunk 3 above)
Comment (minotogna) [reply src/meta/api/request/geo/layer.ts:15]: this is a really good pr, but i share the concern about any. is it possible to find a solution to not use any here?

Comment (minotogna) [reply src/meta/api/request/geo/layer.ts:15]: @sorja we could leave any. it requires to update ```interface ParsedQs { [key: string]: undefined | string | string[] | ParsedQs | ParsedQs[] | number }``` from the query string namespace. 
or do you think we should try to give it a shoot anyway ?

Comment (eraviolo) [reply src/meta/api/request/geo/layer.ts:15]: I get a compilation error if typing ReqQuery with other than string or any:
Screenshot from 2022-11-16 13-47-00

"No overload matches this call.
  The last overload gave the following error.
    Argument of type '(req: ForestLayerRequest, res: Response) => Promise<void>' is not assignable to parameter of type 'RequestHandlerParams<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'"

...because this is the interface:
 interface ParsedQs { [key: string]: undefined | string | string[] | ParsedQs | ParsedQs[] }

Comment (sorja) [reply src/meta/api/request/geo/layer.ts:15]: Scope of this PR any is ok

**Review by minotogna — APPROVED** [review]

**Review by sorja — APPROVED** [review]
